File: source/chatbot/chatbot.js
const tf = require('@tensorflow/tfjs-node');
const fs = require('fs');

// Rutas de los archivos
const modelPath = './modelo_nodejs/model.json'; // Ruta al modelo convertido a TensorFlow.js
const dataPath = './data.json'; // Ruta al archivo de datos con intenciones y respuestas

// Variables para el modelo y los datos
let model;
let intentsData;

// Cargar el modelo TensorFlow.js
async function loadModel() {
    try {
        model = await tf.loadLayersModel(`file://${modelPath}`);
        console.log('Modelo cargado exitosamente.');
    } catch (err) {
        console.error('Error al cargar el modelo:', err);
    }
}

// Cargar datos de intenciones
function loadData() {
    try {
        const rawData = fs.readFileSync(dataPath, 'utf-8');
        intentsData = JSON.parse(rawData);
        console.log('Datos de intenciones cargados.');
    } catch (err) {
        console.error('Error al cargar los datos de intenciones:', err);
    }
}

// Procesar entrada del usuario (dummy tokenizer)
function preprocessInput(userInput, maxLen = 10) {
    // Simulación básica de tokenización (convertir palabras a valores numéricos)
    const words = userInput.toLowerCase().split(' ');
    const wordIndices = words.map((word) => word.charCodeAt(0) % 100); // Convertir caracteres a índices
    const padded = Array(maxLen).fill(0); // Rellenar con ceros
    for (let i = 0; i < Math.min(words.length, maxLen); i++) {
        padded[i] = wordIndices[i];
    }
    return tf.tensor2d([padded]);
}

// Predecir la intención basada en la entrada del usuario
async function predictIntent(userInput) {
    if (!model) {
        throw new Error('El modelo no está cargado.');
    }

    try {
        const inputTensor = preprocessInput(userInput);
        const prediction = model.predict(inputTensor);
        const predictedIndex = prediction.argMax(-1).dataSync()[0]; // Obtener el índice de mayor probabilidad
        return intentsData.intents[predictedIndex].intent;
    } catch (err) {
        console.error('Error al predecir la intención:', err);
        return 'Lo siento, hubo un problema al procesar tu solicitud.';
    }
}

// Obtener una respuesta basada en la intención
function getResponse(intent) {
    const intentObj = intentsData.intents.find((item) => item.intent === intent);
    if (intentObj) {
        const responses = intentObj.responses;
        return responses[Math.floor(Math.random() * responses.length)];
    }
    return 'Lo siento, no entendí eso.';
}

// Inicializar el chatbot
async function chatbot() {
    await loadModel();
    loadData();

    console.log('Chatbot: ¡Hola! Escribe algo para comenzar (o "salir" para terminar).');
    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    rl.on('line', async (userInput) => {
        if (userInput.toLowerCase() === 'salir') {
            console.log('Chatbot: ¡Hasta luego!');
            rl.close();
        } else {
            try {
                const intent = await predictIntent(userInput);
                const response = getResponse(intent);
                console.log(`Chatbot: ${response}`);
            } catch (err) {
                console.error('Error:', err.message);
            }
        }
    });
}

// Ejecutar el chatbot
chatbot();

File: Fase3/chatbot.py
import json
import pickle
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Cargar modelo y recursos
model = tf.keras.models.load_model("modelo_python/chatbot_model.h5")

with open("modelo_python/tokenizer.pkl", "rb") as tokenizer_file:
    tokenizer = pickle.load(tokenizer_file)

with open("modelo_python/label_encoder.pkl", "rb") as encoder_file:
    label_encoder = pickle.load(encoder_file)

# Función para predecir la intención
def predict_intent(user_input):
    # Preprocesar la entrada del usuario
    sequences = tokenizer.texts_to_sequences([user_input])
    padded_sequences = pad_sequences(sequences, maxlen=10)

    # Realizar predicción
    prediction = model.predict(padded_sequences)
    intent_index = np.argmax(prediction)
    intent = label_encoder.inverse_transform([intent_index])[0]

    return intent

# Función para responder al usuario
def get_response(intent):
    with open("datax.json", "r", encoding="utf-8") as file:
        data = json.load(file)
    for item in data["intents"]:
        if item["intent"] == intent:
            return np.random.choice(item["responses"])

# Ciclo principal del chatbot
def chatbot():
    print("Chatbot: ¡Hola! Escribe algo para comenzar (o escribe 'salir' para terminar).")
    while True:
        user_input = input("Tú: ")
        if user_input.lower() == "salir":
            print("Chatbot: ¡Hasta luego!")
            break
        intent = predict_intent(user_input)
        response = get_response(intent)
        print(f"Chatbot: {response}")

# Ejecutar el chatbot
if __name__ == "__main__":
    chatbot()